### PR TITLE
ENH: explicit filters parameter in pd.read_parquet

### LINKED ISSUE #52238: DOC: Document the filters argument in read_parquet
### Pandas version checks

- [X] I have checked that the issue still exists on the latest versions of the docs on `main` [here](https://pandas.pydata.org/docs/dev/)

### Location of the documentation

https://pandas.pydata.org/docs/reference/api/pandas.read_parquet.html

### Documentation problem

The `filters` argument is massively important when reading Parquet files, but it's currently undocumented.  It is documented in the Dask documentation: https://docs.dask.org/en/stable/generated/dask.dataframe.read_parquet.html

I actually didn't even think pandas supported this argument, but apparently, it's supported & undocumented.

### Suggested fix for documentation

I think the Dask documentation can carry over pretty well to the pandas documentation, but we should use language that's easier to understand.  I don't think we should use the "disjunctive normal form (DNF)" terminology - that's just unnecessary.

Let me know if the community supports this fix and I'd be happy to draft some language.

### PR DESCRIPTION
**Pull request by mrastgoo**
- [x] closes #52238 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

added filters as a new parameter in pd.read_parquet

### CODE REVIEW

**Review by mroeschke — CHANGES_REQUESTED** [review]
Instead since we still support both engines, I think linking to the docs for both fastparquet and pyarrow in `**kwargs` is sufficient https://github.com/pandas-dev/pandas/issues/52238#issuecomment-1485571801

**Comment by jorisvandenbossche** [comment]
Given that the `filters` keyword is supported by both engines and works the same, I think it would certainly be useful to explicitly document this keyword, and not hide it in the description of the kwargs. We actually already do the same for the `columns` keyword.

**Comment by jorisvandenbossche** [comment]
@mroeschke OK with adding this?

**Comment by mroeschke** [comment]
Okay yeah that makes sense to add this then

**Review by mroeschke — COMMENTED** [review]
Could you add a test for `filters` for both engines?

**Comment by jorisvandenbossche** [comment]
@mrastgoo would you have time to add a test?

We have an existing test for the columns keyword (`test_read_columns`), so could add an equivalent `test_read_filters` test case next to it. Something like:

```
    def test_read_filters(self, engine, tmp_path):
        df = pd.DataFrame({"int": list(range(4)), "part": list("aabb"),})

        expected = pd.DataFrame({"int": [0, 1]})
        check_round_trip(
            df,
            engine,
            path=tmp_path,
            expected=expected,
            write_kwargs={"partition_cols": ["part"]},
            read_kwargs={"filters": [("string", "==", "a")], "columns":["int"]},
            repeat=1,
        )
```

I used partition cols to actually see the effect of the filter also for fastparquet (to ensure it is correctly passed through). And with that, the `repeat=1` is needed to not add additional files to the directory when writing a second time (with partition_cols, it does not just overwrite the single file).

**Comment by mrastgoo** [comment]
@jorisvandenbossche , yes I would do it asap, sorry for the delay in this issue and thanks for the info I will take that into account.

**Comment by mrastgoo** [comment]
I just pushed for the test. 
I changed the following line for the test
```
read_kwargs={"filters": [("part", "==", "a")], "columns":["int"]},
```
the error which was raised was not very clear for missing columns, which is not a pandas issue, but thought to mention it. 
```
pyarrow.lib.ArrowInvalid: No match for FieldRef.Name(string) in int: int64
part: dictionary<values=string, indices=int32, ordered=0>
```

**Comment by mroeschke** [comment]
Could you merge in main once more?

File: pandas/io/parquet.py [hunk 1]
```diff
@@ -483,6 +489,7 @@ def read_parquet(
     path: FilePath | ReadBuffer[bytes],
     engine: str = "auto",
     columns: list[str] | None = None,
+    filters: list[tuple] | list[list[tuple]] | None = None,
```
Comment (mroeschke) [inline pandas/io/parquet.py:492]: Could you put this as the last argument? (before `**kwargs`)?

File: pandas/io/parquet.py [hunk 2]
```diff
@@ -517,6 +524,21 @@ def read_parquet(
...
+        through an `AND` operation.
+        The outer list combines these sets of filters through an `OR`
+        operation.
+        A single list of tuples can also be used, meaning that no `OR`
+        operation between set of filters is to be conducted.
+
+        Using this argument will NOT result in row-wise filtering of the final
+        partitions unless ``engine="pyarrow"`` is also specified.  For
+        other engines, filtering is only performed at the partition level, that is,
+        to prevent the loading of some row-groups and/or files.
```
Comment (mroeschke) [inline pandas/io/parquet.py:541]: Could you add a `.. versionadded:: 2.1.0` at the end?

**Review by mroeschke — COMMENTED** [review]
Also needs a whatsnew entry in `2.1.0.rst` in the IO section

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @mrastgoo

**Pull request merged**